BUG - remove scaling multiplier from `Period` diff result

ArtinSarraf opened the pull request:
- [ ] closes #23878
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Commit:
BUG/TST - do not multiply period diff result by freq scaling factor

Commit:
TST - reference issue number in test

pep8speaks commented:
Hello @ArtinSarraf! Thanks for submitting the PR.

 - In the file [`pandas/tests/arithmetic/test_period.py`](https://github.com/ArtinSarraf/pandas/blob/b9fe48a9a9f15aef3cbd50dc2b41051090d7e037/pandas/tests/arithmetic/test_period.py), following                are the PEP8 issues :
> [Line 1102:9](https://github.com/ArtinSarraf/pandas/blob/b9fe48a9a9f15aef3cbd50dc2b41051090d7e037/pandas/tests/arithmetic/test_period.py#L1102): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 1108:1](https://github.com/ArtinSarraf/pandas/blob/b9fe48a9a9f15aef3cbd50dc2b41051090d7e037/pandas/tests/arithmetic/test_period.py#L1108): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

Commit:
CLN - pep8 adherence

gfyoung commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -1085,3 +1086,21 @@ def test_pi_sub_period_nat(self):
+    ])
+    def test_period_diff(self, freq, expected):
+        # GH 23878
+        for i in range(1, 4):
```
Parameterize on `i` as well.

ArtinSarraf replied:
Will do.

gfyoung commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -608,8 +608,9 @@ def test_pi_add_offset_n_gt1_not_divisible(self, box_with_array):
-        tm.assert_equal(result, expected)
+        # This test is broken
+        # result = to_offset('3M') + pi
+        # tm.assert_equal(result, expected)
```
Why are we commenting this out?

Existing tests shouldn't be broken unless we have a very good reason for this...

ArtinSarraf replied:
Ah yea, sorry meant to make a comment addressing this in the PR discussion. This was failing for me on a clean checkout of the latest master code, before I made my changes. I can provide more details tomorrow.

gfyoung replied:
If you could revert the change here and let CI run on it, that would be great actually.

gfyoung reviewed (changes requested):
@ArtinSarraf : Good start!  I have a couple of comments regarding the tests. 

Also, don't forget to add a `whatsnew` for this bug.

mroeschke commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -608,8 +608,9 @@ def test_pi_add_offset_n_gt1_not_divisible(self, box_with_array):
 
-        result = to_offset('3M') + pi
-        tm.assert_equal(result, expected)
+        # This test is broken
```
Why is this test broken?

Nonetheless, we'd prefer to xfail if it's supposed to fail instead of commenting this out.

gfyoung replied:
@mroeschke : Perhaps it's lag, but you're repeating my comments from above.

mroeschke commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -1085,3 +1086,21 @@ def test_pi_sub_period_nat(self):
+    ])
+    def test_period_diff(self, freq, expected):
+        # GH 23878
+        for i in range(1, 4):
```
you can `pytest.mark.parametrize` over this as well.

gfyoung replied:
@mroeschke : Perhaps it's lag, but you're repeating my comments from above.

ArtinSarraf commented:
@gfyoung - the reason I didn’t add a whatsnew entry is that this Offset result from diffing periods is already new behavior in 0.24 so the bug had not been released yet. Should I still add an entry for it?

gfyoung commented:
> the reason I didn’t add a whatsnew entry is that this Offset result from diffing periods is already new behavior in 0.24 so the bug had not been released yet. Should I still add an entry for it?

Ah, gotcha.  In that case, why don't you add a reference to your issue number to the existing `whatsnew` entry instead.

Commit:
DOC - reference issue in whatsnew

Commit:
Revert "DOC - reference issue in whatsnew"

This reverts commit 364d4a99ecc72569973e35621fa83fafbd517ac8.

Commit:
TST - parameterize test

Commit:
DOC - reference issue in whatsnew

jbrockmendel commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1685,7 +1685,7 @@ cdef class _Period(object):
                     msg = _DIFFERENT_FREQ.format(self.freqstr, other.freqstr)
                     raise IncompatibleFrequency(msg)
-                return (self.ordinal - other.ordinal) * self.freq
+                return (self.ordinal - other.ordinal) * type(self.freq)()
```
Won't this be wrong for offsets that have relevant keywords?

ArtinSarraf replied:
You’re right. So one option would be to do `... * self.freq * (1. / self.freq.n)`. (Since DateOffset doesn’t support div) This way no other classes need to be modified. However, this could probably lead to unforeseen floating point issues, so I think it might be worth adding a property to the DateOffset objet to provide the object with n=1. Something like `DateOffset.base_freq`? Or __div__ that only allows factors of n.

ArtinSarraf replied:
done

jbrockmendel commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -1085,3 +1085,21 @@ def test_pi_sub_period_nat(self):
         tm.assert_index_equal(pd.Period('NaT', freq='M') - idx, exp)
+
+
+class TestPeriodArithmetic(object):
```
This all belongs in pandas.tests.scalar.period.  If you want to make a new file test_arithmetic.py in that directory, that'd be OK.  Otherwise put in test_period.py

ArtinSarraf replied:
moved

jbrockmendel commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -1085,3 +1085,21 @@ def test_pi_sub_period_nat(self):
+        (pd.offsets.Day, 214),
+        (pd.offsets.MonthEnd, 7),
+        (pd.offsets.YearEnd, 1),
+    ])
```
definitely needs cases with kwargs passed to offset constructors.  Putting it in tests.tseries.offsets might be useful since there are test classes that construct a bunch of these

ArtinSarraf replied:
parameterized the tests on some kwargs too, only YearEnd takes any kwds besides normalize ('month'). And only non-Tick offsets (in this case MonthEnd and YearEnd) can take normalize=True.

Commit:
BUG - account for freq kwds

Commit:
TST - move non standard freq period diff test and test various keywords

ArtinSarraf commented:
@gfyoung / @jbrockmendel - any other changes to consider?

gfyoung commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1056,7 +1056,7 @@ from :class:`Period`, :class:`PeriodIndex`, and in some cases
 
 This usage is now deprecated.  Instead add or subtract integer multiples of
-the object's ``freq`` attribute (:issue:`21939`)
+the object's ``freq`` attribute (:issue:`21939`) / (:issue:`23878`)
```
(:issue:`21939`) / (:issue:`23878`) --> (:issue:`21939`, :issue:`23878`)

ArtinSarraf replied:
done

jbrockmendel commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1685,7 +1685,9 @@ cdef class _Period(object):
                     raise IncompatibleFrequency(msg)
-                return (self.ordinal - other.ordinal) * self.freq
+                base_freq = type(self.freq)(normalize=self.freq.normalize,
+                                            **self.freq.kwds)
```
pass `n=1` explicitly here.

add a reference `# GH#23915` for future readers

ArtinSarraf replied:
done

Commit:
BUG/CLN - fix periodindex/array diff, and provide base method for offsets

Commit:
TST/DOC - split tick and offset tests and fix whatsnew

Commit:
TST - add periodindex diff fix tests

jreback commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -352,6 +352,10 @@ class _BaseOffset(object):
 
+    @property
+    def base(self):
+        return type(self)(n=1, normalize=self.normalize, **self.kwds)
```
can you add a doc-string here

ArtinSarraf replied:
added.

ArtinSarraf commented on pandas/tests/scalar/period/test_period.py:
```diff
@@ -1106,6 +1106,38 @@ def test_sub(self):
+                expected = 0
+            else:
+                return
+        # Only non-Tick frequencies can have normalize set to True
```
Separated.
Only 4 of the non-tick frequencies in pd.offsets are valid Period frequencies so I explicitly parameterized those in the tests. The Tick fixture worked well though.

ArtinSarraf commented:
@jbrockmendel - added the fix for the PeriodIndex diff as well and added tests. Also added a `base` method to BaseOffset to get the n=1 frequency of a given offset object, since it was getting used a few times in the tests and source.

Commit:
DOC - add docstrings and GH references

jreback commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1685,7 +1685,8 @@ cdef class _Period(object):
                     msg = _DIFFERENT_FREQ.format(self.freqstr, other.freqstr)
                     raise IncompatibleFrequency(msg)
-                return (self.ordinal - other.ordinal) * self.freq
+                # GH 23915
```
can you add a 1-line comment on what this is needed / doing

jreback reviewed (approved):
minor comment, ping on green.

jreback commented:
@gfyoung if you have any comments.

jreback commented:
@jbrockmendel if you'd have a look at the tests and see if they are sufficient

Commit:
DOC - additional explanation of code

jreback commented:
@ArtinSarraf can you merge master and see if you can resolve failures

Commit:
Merge branch 'master' into bug/period_diff

ArtinSarraf commented:
@jreback / @jbrockmendel - is there any way to recreate the testing envs of the automated tests. My tests run fine locally and from the failed test output its not obvious what the error is (since the repr of the result and expected are the same), I'm assuming either the normalize kwd attributes are differing somehow.

Commit:
Merge branch 'master' into bug/period_diff

Commit:
CLN - reorganize test

ArtinSarraf commented:
@jreback / @jbrockmendel

Found the issue. Looks like this is due to an existing bug with PeriodIndex (I will open a separate issue for this). See the example below.

    >>> pd.PeriodIndex(['19910905'], freq=pd.offsets.YearEnd(normalize=True)).freq.normalize  # -> True
    >>> pd.PeriodIndex(['19910905'], freq=pd.offsets.YearEnd(normalize=False)).freq.normalize  # -> True

Restart the process

    >>> pd.PeriodIndex(['19910905'], freq=pd.offsets.YearEnd(normalize=False)).freq.normalize  # -> False
    >>> pd.PeriodIndex(['19910905'], freq=pd.offsets.YearEnd(normalize=True)).freq.normalize  # -> False

Looks like the `normalize` option gets cached for PeriodIndex somehow. This causes my tests to fail because I iterate through `normalize = True | False`.

jbrockmendel commented:
> Looks like the normalize option gets cached for PeriodIndex somehow

Good catch.  Best guess is it is in pandas.tseries.frequencies

Commit:
TST - update tests to account for existing bug

ArtinSarraf commented:
@jbrockmendel - looks like the only failing test in the pandas-dev.pandas tests are linting errors now due to importing the tick_classes fixture (since it's not in the discovery path for those tests).

    from pandas.tests.tseries.offsets.conftest import tick_classes

Is there another way to discover this fixture? I could move the test, but I think it makes sense to keep it grouped where it is.

jreback commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -16,6 +16,7 @@
 from pandas.core import ops
 from pandas import Period, PeriodIndex, period_range, Series
 from pandas.tseries.frequencies import to_offset
```
so we can't do this, instead move the tick_classes fixture to pandas/conftest.py. I *think* everything should still work

ArtinSarraf replied:
done.

Commit:
TST - move tick fixture to pandas/conftest

Commit:
CLN - fix linting error

jreback commented:
@ArtinSarraf lgtm. can you add a whatsnew note. ping on green.

@gfyoung good?

Commit:
DOC - add more detail about this fixs behavior in the whatsnew

Commit:
DOC - fix issue reference

Commit:
DOC - fix spacing

ArtinSarraf commented:
@jreback - tests are all clean.

jreback commented:
thanks @ArtinSarraf
